New whitening scheme based on inverse spectrum truncation

alurban opened the pull request:
This pull request addresses a common issue with filter transients in the current whitening method by replacing it with an alternative scheme which designs and applies an FIR (finite impulse response) filter using inverse spectrum truncation.

Other new features added include:
* A method `FrequencySeries.interpolate()` and functions `timeseries._fft_length_default`, `signal.filter_design.truncate_transfer`, and `signal.filter_design.truncate_impulse`, all used to design the FIR filter
* New keyword arguments for filter duration (in the time domain) and highpass corner frequency
* In `TimeSeries.whiten()`, a default value of `fftlength`which  causes it to be set to `max(2, int(2048 // self.sample_rate.decompose().value))`
* Automatic cropping of the beginning and end of the whitened timeseries by a segment of length `filter_duration`
* Updates to the documentation for `TimeSeries.whiten()` addressing these changes
* Updates to `q_transform()` that address these recent changes, including a new default `fftlength` that mirrors the one for `TimeSeries.whiten`
* Unit tests for new functionality, a unit test for `signal.window.planck()`, and a bug fix in the unit test for `FrequencySeries.inject()`

Note, these changes make it so that a simple call to `TimeSeries.whiten()` with no other arguments returns something sensible in most cases. It is also backwards compatible.

This fixes #867. It is also related to #358 and #451.

alurban commented:
@duncanmmacleod, I've compared the whitening schemes by plotting the timeseries in figure 1 (original scheme in orange, new scheme in blue). In figure 2 the difference between the two is shown in blue, while the bandpassed difference is shown in orange to demonstrate that the filter artifacts are not produced by the new scheme. (You can see some 1 Hz beat notes / windowing artifacts in the orange timeseries.)

![170817_compare_whitening](https://user-images.githubusercontent.com/5273800/44010173-be6001ec-9e65-11e8-8d47-796f34a38df8.png)
![170817_whitening_difference](https://user-images.githubusercontent.com/5273800/44010176-c56bb3dc-9e65-11e8-9214-c38f7f8ae27e.png)


On a separate note, it looks like there are a few issues with this PR. I'll have a look at those in short order. One issue is python 3 readiness, which I think i can resolve with modest effort. The other issues are from `codeclimate` and related to complexity. There's not much I can do about this as the arguments I've added are all necessary, except perhaps the `highpass` argument, and the large complexity and number of lines come from supporting two separate whitening schemes. I didn't want to delete your existing scheme, presumably you want to keep it in there? Do you have any thoughts?

For my purposes, this alternative whitening scheme would be very nice to have in 0.12.0 as I want to use it for `gwdetchar-omega`.

duncanmmacleod commented on gwpy/timeseries/timeseries.py:
```diff
@@ -1428,14 +1428,15 @@ def taper(self, side='leftright'):
-        """White this `TimeSeries` against its own ASD
+    def whiten(self, fftlength=2, overlap=0, method='scipy-welch',
+               window='hanning', detrend='constant', asd=None,
+               fir=False, filter_duration=1, highpass=0, **kwargs):
```
I think I would be more comfortable with `highpass=None` as the default, to make it more obvious that the default is to not apply a FD highpass.

alurban replied:
Makes sense to me, and that's a pretty easy change, so I'll make that the default.

duncanmmacleod commented on gwpy/timeseries/timeseries.py:
```diff
@@ -1428,14 +1428,15 @@ def taper(self, side='leftright'):
-        """White this `TimeSeries` against its own ASD
+    def whiten(self, fftlength=2, overlap=0, method='scipy-welch',
+               window='hanning', detrend='constant', asd=None,
+               fir=False, filter_duration=1, highpass=0, **kwargs):
```
is it reasonable to tie the `filter_duration` to `fftlength` and remove the former option? they seem redundant to me, but I understand they are currently used in separate contexts.

alurban replied:
I prefer to keep `fftlength` and `filter_duration` as separate options. In this new scheme, the former is a feature of the ASD calculation, while the latter is a feature of the actual filter that doesn't necessarily relate to `fftlength` at all. This is one reason why I've introduced a `FrequencySeries.interpolate` method, which is very similar to what [PyCBC](http://pycbc.org/pycbc/latest/html/_modules/pycbc/psd/estimate.html#interpolate) does.

duncanmmacleod replied:
Ok, sounds reasonable.

duncanmmacleod commented on gwpy/timeseries/timeseries.py:
```diff
@@ -1428,14 +1428,15 @@ def taper(self, side='leftright'):
-    def whiten(self, fftlength, overlap=0, method='scipy-welch',
-               window='hanning', detrend='constant', asd=None, **kwargs):
-        """White this `TimeSeries` against its own ASD
+    def whiten(self, fftlength=2, overlap=0, method='scipy-welch',
```
why was a default added here? and why this value?

alurban replied:
I added the default because under the new scheme, `fftlength` isn't used at all if `fir=True` and `asd` is not `None`. I chose a default of 2 seconds because that's the smallest use case I see for this parameter, and I see it quite often.

duncanmmacleod commented on gwpy/frequencyseries/frequencyseries.py:
```diff
@@ -269,6 +270,37 @@ def zpk(self, zeros, poles, gain, analog=True):
+        ----------
+        df : `float`
+            desired frequency resolution of the interpolated `FrequencySeries`,
+            in Hz
```
can you leave a blank line between parameter definitions in the docstring, it makes things a little easier to read IMO

alurban replied:
Sure, I'll make this change as well

duncanmmacleod commented on gwpy/timeseries/timeseries.py:
```diff
@@ -1476,11 +1488,48 @@ def whiten(self, fftlength, overlap=0, method='scipy-welch',
+            in_ = self.detrend(detrend)
+            out = numpy.zeros(self.size).view(type(self))
+            out.__metadata_finalize__(self)
+            out._unit = self.unit
```
you should be able to use `out.__array_finalize__` to set the units and other metadata in one call, I haven't updated other methods to do this yet.

duncanmmacleod commented on gwpy/timeseries/timeseries.py:
```diff
@@ -1476,11 +1488,48 @@ def whiten(self, fftlength, overlap=0, method='scipy-welch',
+            out.__metadata_finalize__(self)
+            out._unit = self.unit
+            out.value[:] += npfft.irfft(in_.fft().value * invasd)
+            out.value[:] *= self.duration.value / 2  # re-normalise
```
please combine these into a single operation (unless there's a good reason not to)

duncanmmacleod commented on gwpy/timeseries/timeseries.py:
```diff
@@ -1476,11 +1488,48 @@ def whiten(self, fftlength, overlap=0, method='scipy-welch',
+            # apply truncated filter in the frequency domain
+            invasd = numpy.abs(npfft.rfft(tdw))  # enforce zero-phase
+            in_ = self.detrend(detrend)
+            out = numpy.zeros(self.size).view(type(self))
```
it seems cleaner to just do something like

```python
out = type(self)(npfft.irfft(in_.fft().value * invasd) * self.duration.value/2)
out.__array_finalize__(self)
```

duncanmmacleod commented on gwpy/timeseries/timeseries.py:
```diff
@@ -1476,11 +1488,48 @@ def whiten(self, fftlength, overlap=0, method='scipy-welch',
+            out._unit = self.unit
+            out.value[:] += npfft.irfft(in_.fft().value * invasd)
+            out.value[:] *= self.duration.value / 2  # re-normalise
+            del tdw, invasd, in_, out.times
```
why are these `del` required?

alurban replied:
I think I was emulating what's done in another method, can't remember which one. Are those arrays deleted from memory automatically after the function returns?

duncanmmacleod replied:
As soon as there are no references to them, the garbage collector will remove them automatically, so as long as they aren't linked to from the returned `out` array, this should not be necessary.

alurban commented:
@duncanmmacleod, I've addressed your review comments and removed the old whitening scheme entirely. The new whitening method has the same name (`TimeSeries.whiten()`) and same keyword arguments, with two new ones for filter duration and highpass. This should satisfy the requirement of being backward compatible.

However, now there are two build errors in `test_q_transform`. I'll have a look at that shortly.

alurban commented:
@duncanmmacleod, I've resolved WIP status and am ready for review. I fixed the errors by automatically cropping filter corruptions out of the whitened timeseries, then adjusting the `q_transform` and various unit tests to account for this.

alurban commented:
@duncanmmacleod, for illustration, here are the whitened timeseries (figure 1) and a Q-spectrogram (figure 2) of `L1:GDS-CALIB_STRAIN` at the end of GW170817. I made these using the most recent version of the new whitening method with `fftlength=8`, `overlap=4`, `qrange=(93.1, 93.1)`, and all other options default.

![170817_whitened_timeseries](https://user-images.githubusercontent.com/5273800/44072859-e0e9b722-9f45-11e8-8f39-a867eed7f26e.png)
![170817_strain](https://user-images.githubusercontent.com/5273800/44072870-e7408aec-9f45-11e8-8af0-c8679a62b1e9.png)

I'm quite happy with this, even with a 1-second filter the glitch is well-resolved and doesn't really ring off or corrupt the filter much at all. You can even see where the glitch isn't quite loud enough to surpass a few lines in the spectrum, and the BNS trace is clearly visible as well.

After I have some ice cream, I'll try squashing these commits.

duncanmmacleod commented on gwpy/timeseries/timeseries.py:
```diff
@@ -1428,14 +1428,17 @@ def taper(self, side='leftright'):
+               window='hanning', detrend='constant', asd=None,
+               fduration=1, highpass=None, **kwargs):
+        """Whiten this `TimeSeries` against its own ASD using inverse spectrum
+           truncation
```
Can you shorten this docstring to

```
Whiten this `TimeSeries` using inverse spectrum truncation
```

? This will make it fit on one line, and the whitening doesn't have to use 'its own ASD'.

alurban replied:
This is fixed in commit 66e9982, along with a pesky split infinitive elsewhere in that doctoring.

duncanmmacleod commented on gwpy/timeseries/timeseries.py:
```diff
@@ -1476,36 +1489,43 @@ def whiten(self, fftlength, overlap=0, method='scipy-welch',
+        out = type(self)(npfft.irfft(in_.fft().value * invasd))
+        out.__array_finalize__(self)
+        return out.crop(out.t0.decompose().value + fduration,
+                        (out.t0 + out.duration).decompose().value - fduration)
```
it would be cleaner here to use `out.xspan`, which does all decompositions internally, something like

```python
return out.crop(*out.xspan.contract(fduration))
```

alurban replied:
I like that a lot better as well! This is fixed in commit 66e9982.

pep8speaks commented:
Hello @alurban! Thanks for updating the PR.











 - In the file [`gwpy/signal/tests/test_window.py`](https://github.com/gwpy/gwpy/blob/8f64454ab97da67a2a2d8f06ab3da442edf9d120/gwpy/signal/tests/test_window.py), following are the PEP8 issues :

> [Line 48:1](https://github.com/gwpy/gwpy/blob/8f64454ab97da67a2a2d8f06ab3da442edf9d120/gwpy/signal/tests/test_window.py#L48): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1



 - In the file [`gwpy/timeseries/timeseries.py`](https://github.com/gwpy/gwpy/blob/8f64454ab97da67a2a2d8f06ab3da442edf9d120/gwpy/timeseries/timeseries.py), following are the PEP8 issues :

> [Line 54:1](https://github.com/gwpy/gwpy/blob/8f64454ab97da67a2a2d8f06ab3da442edf9d120/gwpy/timeseries/timeseries.py#L54): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 1529:39](https://github.com/gwpy/gwpy/blob/8f64454ab97da67a2a2d8f06ab3da442edf9d120/gwpy/timeseries/timeseries.py#L1529): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent

alurban commented:
@duncanmmacleod, I think we're ready for primetime. I've refactored `TimeSeries.whiten()` such that the filter truncation is done with two new functions, `signal.filter_design.truncate_transfer` and `signal.filter_design.truncate_impulse` which smoothly truncate (and, optionally, highpass) the transfer function and impulse response of the desired filter. I've also added a global function, `timeseries._fft_length_default`, which sets the default FFT length and is invoked by `TimeSeries.whiten()` and `TimeSeries.q_transform()`. Unit tests for the new filter design functions are also included.

On a separate note, I managed to reproduce the original Q-spectrogram for GW150914 by using a filter duration of 2 seconds, the same as the FFT length:

![150914_strain](https://user-images.githubusercontent.com/5273800/44117140-4539428e-9fc7-11e8-85d8-9dd0a086fcc6.png)

I still think the default filter length of 1 second is a sensible default though, and I did use it to reproduce the Q-spectrogram for GW170817 above.

duncanmmacleod commented on gwpy/frequencyseries/frequencyseries.py:
```diff
@@ -269,6 +270,38 @@ def zpk(self, zeros, poles, gain, analog=True):
+
+        See Also
+        --------
+        :func:`numpy.interp` for the underlying 1-D linear interpolation scheme
```
This usage of `see also` doesn't work, numpydoc expects the format to be:

```
See Also
--------
:func:`numpy.interp`
    for the undelying 1-D linear interpolation scheme
```
I think. The new [numpydoc docs](https://numpydoc.readthedocs.io/en/latest/format.html#sections) suggest that you might be able to use a simple colon-separated key, value listing, but I haven't tested that.

alurban replied:
Good catch! This is addressed in commit 58c95a9.

duncanmmacleod commented:
Thanks @alurban!
